openssl/src/x509/sbgp.rs

```diff
@@ -0,0 +1,232 @@
+#[cfg(ossl110)]
+use std::convert::TryInto;
+#[cfg(ossl110)]
+use std::mem::MaybeUninit;
+
+#[cfg(ossl110)]
+use ffi::{
+    ASIdOrRange_id, ASIdOrRange_range, ASIdentifierChoice_asIdsOrRanges,
+    ASIdentifierChoice_inherit, IPAddressChoice_addressesOrRanges, X509v3_addr_get_afi,
+    X509v3_addr_get_range, ASN1_INTEGER, IANA_AFI_IPV4, IANA_AFI_IPV6,
+};
+#[cfg(ossl110)]
+use foreign_types::{ForeignType, ForeignTypeRef};
+
+#[cfg(ossl110)]
+use crate::{
+    asn1::Asn1IntegerRef,
+    stack::{Stack, StackRef, Stackable},
+    util::{ForeignTypeExt, ForeignTypeRefExt},
+};
+
+#[cfg(ossl110)]
+use super::X509;
+
+#[cfg(ossl110)]
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::ASIdOrRange;
+    fn drop = ffi::ASIdOrRange_free;
+
+    /// The AS number extension of an `X509` certificate.
+    pub struct ASIdOrRange;
+    /// Reference to `ASIdOrRange`.
+    pub struct ASIdOrRangeRef;
+}
+#[cfg(ossl110)]
+impl Stackable for ASIdOrRange {
+    type StackType = ffi::stack_st_ASIdOrRange;
+}
+
+#[cfg(ossl110)]
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::ASIdentifiers;
+    fn drop = ffi::ASIdentifiers_free;
+
+    /// The AS number extension of an `X509` certificate.
+    pub struct ASIdentifiers;
+    /// Reference to `ASIdentifiers`.
+    pub struct ASIdentifiersRef;
+}
+
+#[cfg(ossl110)]
+impl ASIdentifiers {
+    pub fn inherited(&self) -> bool {
+        unsafe {
+            let asptr = self.0;
+            let asnum = (*asptr).asnum;
+            (*asnum).type_ == ASIdentifierChoice_inherit
+        }
+    }
+
+    pub fn ranges(&self) -> Option<Vec<(u32, u32)>> {
```

Comment:
Does this correspond to some OpenSSL function?

Reply:
Same question for the other methods in this file.

Reply:
These function correspond to `X509v3_addr_get_range`, but executed for the whole extension, since RFC 3779 encodings are rather tedious to work with. Same thing applies to the other functions in this file, they just parse the openssl-internal representation into their Rust-equivalents.